lib/rspec/support/object_inspector.rb

```diff
@@ -4,19 +4,39 @@ module Support
     # printing Time, DateTime, or BigDecimal
     module ObjectInspector
       # @api private
-      def self.inspect(object)
-        if Time === object
+      # rubocop:disable CyclomaticComplexity
+      # rubocop:disable MethodLength
+      def self.inspect(object, nested=false)
+        case
+        when Time === object
           format_time(object)
-        elsif defined?(DateTime) && DateTime === object
+        when defined?(DateTime) && DateTime === object
           format_date_time(object)
-        elsif defined?(BigDecimal) && BigDecimal === object
+        when defined?(BigDecimal) && BigDecimal === object
           "#{object.to_s 'F'} (#{object.inspect})"
-        elsif RSpec::Support.is_a_matcher?(object) && object.respond_to?(:description)
+        when defined?(Hash) && Hash === object
+          format_hash(object, nested)
+        when defined?(Array) && Array === object
+          "[#{object.map { |element| ObjectInspector.inspect(element, true) }.join(', ')}]"
+        when RSpec::Support.is_a_matcher?(object) && object.respond_to?(:description)
           object.description
         else
           object.inspect
         end
       end
+      # rubocop:enable CyclomaticComplexity
+      # rubocop:enable MethodLength
+
+      # @private
+      def self.format_hash(object, nested)
+        objects = object.keys.sort_by { |k| k.to_s }.map do |key|
+          formatted_key = ObjectInspector.inspect(key, true)
+          formatted_value = ObjectInspector.inspect(object[key], true)
+          "#{formatted_key} => #{formatted_value}#{',' unless nested}"
+        end
+
+        nested ? "{#{objects.join(', ')}}" : objects.join("\n")
```

Comment:
I noticed that you added some tests of how the nesting interacts with the differ but it would also be nice to see how the nesting affects raw `ObjectInspector.inspect` output since that method is going to be called directly from rspec-expectations and rspec-mocks.